examples/pmatch.kl

```diff
@@ -0,0 +1,145 @@
+(import (shift "free-identifier-case.kl" 1))
+(import (shift "temporaries.kl" 1))
+(import (shift "pair-datatype.kl" 1))
+(import (shift "list-datatype.kl" 1))
+
+(meta
+  (define identifier?
+    (lambda (stx)
+      (syntax-case stx [(ident x) (true)] [_ (false)]))))
+
+(meta
+  (datatype (Maybe A)
+    (nothing)
+    (just A)))
+
+(meta
+  (defun syntax->list (stx)
+    (syntax-case stx
+      [() (nil)]
+      [(cons x xs) (:: x (syntax->list xs))])))
+
+(meta
+  (define else-case
+     (lambda (stx)
+       (syntax-case stx
+         [(list (e x))
+          (if (identifier? e)
+              (free-identifier-case e
+                [else (pure (just x))]
+                [(else other) (pure (nothing))])
+              (pure (nothing)))]
+         [(ident e)
+          (free-identifier-case e
+            [else (pure (just 'else-x))]
+            [(else other) (pure (nothing))])]
+         [_ (pure (nothing))]))))
+
+(meta
+  (defun list->syntax (stx-lst stx)
+    (case stx-lst
+      [(nil) (empty-list-syntax stx)]
+      [(:: x xs) (cons-list-syntax x (list->syntax xs stx) stx)])))
+
+(meta
+  (datatype (Abstracted)
+    (none)
+    (bind Syntax Syntax (Abstracted)))
+
+  (defun abstract (stxs)
+    (syntax-case stxs
+      [() (none)]
+      [(cons e es)
+       (bind 'internal-x e (abstract es))])))
+   
+
+(define-macros
+  ([pmatch
+    (lambda (stx)
+      (syntax-case stx
+        [(cons _ more)
+         (syntax-case more
+           [(cons scrut pats)
+            (pure `(let (x ,scrut) (pmatch-aux ,scrut x ,pats)))])]))]
+   [pmatch-aux
+    (lambda (stx)
+      (syntax-case stx
+        [(list (_ scrut tgt pats))
+         (syntax-case pats
+           [() (pure `(error ',scrut))]
+           [(cons c cs)
+            (syntax-case c
+              [(list (pat rhs))
+               (>>= (else-case pat)
+                    (lambda (e)
+                      (case e
+                        [(just x) (pure `(let (,x ,tgt) ,rhs))]
+                        [(nothing)
+                         (pure `(let (fk (lambda (_) (pmatch-aux ,scrut ,tgt ,cs)))
+                                  (ppat ,tgt ,pat ,rhs (fk 'hi))))])))])])]))]
+   [ppat
+    (lambda (stx)
+      (syntax-case stx
+        [(list (_ tgt pat ks kf))
+         (syntax-case pat
+           [(ident x)
+            (pure `(let (,x ,tgt) ,ks))]
+           [(cons what args)
+            (>>= (make-temporaries (syntax->list args))
+                 (lambda (temps)
+                   (let (temp-names (map fst temps))
+                     (flet (combine (stxs)
+                             (case stxs
+                               [(nil) ks]
+                               [(:: id-and-stx rest)
+                                `(ppat ,(fst id-and-stx) ,(snd id-and-stx) ,(combine rest) ,kf)]))
+                        (pure `(case ,tgt
+                                 [,(cons-list-syntax what (the Syntax (list->syntax temp-names pat)) pat)
```

Comment:
What is the purpose of the `the Syntax` type annotation? The name `list->syntax` already communicates quite clearly that it is returning a `Syntax`. I am guessing that this is leftover from a previous debugging session?

Reply:
On that note, I had completely forgotten that Klister was a typed language while reading this code. How about defining versions of `define`, `let` and `flet` which encourages including a type signature?

Reply:
The annotation is going away - it was for sanity while I was afflicted by #48.

For the `define` etc, we would need a way to annotate type schemes, which we don't yet have. See #60 for a design sketch.

Reply:
Done.